src/rgb_led.rs

```diff
@@ -0,0 +1,230 @@
+use std::{ffi::c_void, thread, time::Duration};
+
+use esp_idf_svc::sys::{
+    esp_err_t, gpio_num_t, rmt_carrier_level_t_RMT_CARRIER_LEVEL_HIGH,
+    rmt_carrier_level_t_RMT_CARRIER_LEVEL_LOW, rmt_channel_t, rmt_config_t,
+    rmt_config_t__bindgen_ty_1, rmt_mode_t_RMT_MODE_TX, rmt_tx_config_t, ESP_OK,
+};
+
+const WS2812_T0H_NS: u32 = 350;
+const WS2812_T0L_NS: u32 = 1000;
+const WS2812_T1H_NS: u32 = 1000;
+const WS2812_T1L_NS: u32 = 350;
+//const WS2812_RESET_US: u32 = 280;
+
+#[derive(Debug)]
+pub struct EspError {
+    inner: esp_err_t,
+}
+
+impl std::fmt::Display for EspError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "esp error code: {}", self.inner)
+    }
+}
+
+impl std::error::Error for EspError {}
+
+fn esp_res(err_code: esp_err_t) -> Result<(), EspError> {
+    if err_code == ESP_OK {
+        Ok(())
+    } else {
+        Err(EspError { inner: err_code })
+    }
+}
+
+pub type Led = LedStrip<1>;
+
+#[derive(Debug)]
+pub struct LedStrip<const NUM_LEDS: usize> {
+    ws2812_t0h_ticks: u32,
+    ws2812_t0l_ticks: u32,
+    ws2812_t1h_ticks: u32,
+    ws2812_t1l_ticks: u32,
+    buffer: [[u8; 3]; NUM_LEDS],
+    channel: rmt_channel_t,
+}
+
+impl<const NUM_LEDS: usize> LedStrip<NUM_LEDS> {
+    pub fn new(rmt_channel: rmt_channel_t, gpio_num: gpio_num_t) -> Result<Self, EspError> {
+        let config = rmt_config_t {
+            rmt_mode: rmt_mode_t_RMT_MODE_TX,
+            gpio_num,
+            channel: rmt_channel,
+            clk_div: 2,
+            mem_block_num: 1,
+            flags: 0,
+            __bindgen_anon_1: rmt_config_t__bindgen_ty_1 {
+                tx_config: rmt_tx_config_t {
+                    carrier_freq_hz: 38000,
+                    carrier_level: rmt_carrier_level_t_RMT_CARRIER_LEVEL_HIGH,
+                    idle_level: rmt_carrier_level_t_RMT_CARRIER_LEVEL_LOW,
+                    carrier_duty_percent: 33,
+                    carrier_en: false,
+                    loop_en: false,
+                    idle_output_en: true,
+                    loop_count: 0,
+                },
+            },
+        };
+
+        let config_ptr: *const rmt_config_t = &config;
+        unsafe {
+            esp_res(esp_idf_svc::sys::rmt_config(config_ptr))?;
+            esp_res(esp_idf_svc::sys::rmt_driver_install(config.channel, 0, 0))?;
+        }
+
+        let mut counter_clk_hz: u32 = 0;
+        unsafe {
+            esp_res(esp_idf_svc::sys::rmt_get_counter_clock(
+                config.channel,
+                &mut counter_clk_hz,
+            ))?;
+        }
+
+        let ratio = counter_clk_hz as f32 / 1e9;
+
+        let ws2812_t0h_ticks = (ratio * WS2812_T0H_NS as f32) as u32;
+        let ws2812_t0l_ticks = (ratio * WS2812_T0L_NS as f32) as u32;
+        let ws2812_t1h_ticks = (ratio * WS2812_T1H_NS as f32) as u32;
+        let ws2812_t1l_ticks = (ratio * WS2812_T1L_NS as f32) as u32;
+
+        unsafe {
```

Comment:
What does unsafe do?

Reply:
[unsafe](https://doc.rust-lang.org/rust-by-example/unsafe.html) lets you bypass some of the compiler's protections. It's never good to have, but it's sometimes necessary. Since Rust cannot guarantee that any C code is safe, any bindings to external C code is unsafe by default.